infra/experimental/sanitizers/ExecSan/execSan.cpp

```diff
@@ -98,24 +180,101 @@ std::vector<std::byte> read_memory(pid_t pid, unsigned long long address, size_t
   return memory;
 }
 
-void inspect(pid_t pid, const user_regs_struct &regs) {
-  auto memory = read_memory(pid, regs.rdi, kTripWire.length());
-  if (memory.size() == 0) {
-    return;
+// Construct a string with the memory specified in a register.
+std::string read_register_string(pid_t pid, unsigned long reg, unsigned long length) {
+  auto memory = read_memory(pid, reg, length);
+  if (!memory.size()) {
+    return "";
   }
 
-  std::string path(reinterpret_cast<char*>(
-        memory.data()), std::min(memory.size(), kTripWire.length()));
+  std::string content(reinterpret_cast<char*>(
+        memory.data()), std::min(memory.size(), length));
+  return content;
+}
+
+void report_bug(int bug_code) {
+  // Report the bug found based on the bug code.
+  fprintf(stderr, "===BUG DETECTED: %s===\n", kBugMessageMap[bug_code].c_str());
+  // Rely on sanitizers/libFuzzer to produce a stacktrace by sending SIGABRT
+  // to the root process.
+  // Note: this may not be reliable or consistent if shell injection happens
```

Comment:
Interesting. How does it happen in an async way?

Reply:
This is from a previous commit. 

If the root process calling the process that causes the injection does not wait for its children, it's possible the stacktrace at this time will not be very consistent.